### DISCUSSION TITLE
Automated generation of feature_levels.yaml file based on feature hierarchy in GFF file

### DISCUSSION THREAD

**Discussion by svigneau** [NBISweden/AGAT#310] [Ideas]
From my experience testing AGAT, it appears critical to have a representation in the feature_levels.yaml file that closely matches the hierarchy in the GFF file being handled, otherwise unintended fixes may be introduced by AGAT. The process of updating the feature_levels.yaml file can be quite laborious when the GFF file contains a lot of feature types with a complex graph.

One such example is the [SGD reference GFF](http://sgd-archive.yeastgenome.org/sequence/S288C_reference/genome_releases/), which has a combination of RNA and non-RNA features, self-referencing feature types, feature types with multiple parent types, or leaf feature types pointing directly to root feature types ([here](https://drive.google.com/file/d/1oItDVksOw0rpq7PIWri_gzdSwveRlRpz/view?usp=sharing) is the corresponding graph).

Do you have a recommended way to automate the process of generating a good feature_levels.yaml file or is it something on the roadmap?

**Comment by Juke34** [comment]
> From my experience testing AGAT, it appears critical to have a representation in the feature_levels.yaml file that closely matches the hierarchy in the GFF file being handled, otherwise unintended fixes may be introduced by AGAT. 

You are definitly right. The features already described in the `feature_levels.yaml` are the most common features and are sufficient in most of cases. Their order (level1,level2,level3) respect the sequence ontology (SO) description, so exepted for specific cases the order should not be modify.

> The process of updating the feature_levels.yaml file can be quite laborious when the GFF file contains a lot of feature types with a complex graph.

When feature types are missing you can add them. I agree that it can be laborious but I didn't find any better solution to make AGAT universal (working with all cases) without any minimal user input. I consider it as semi-automated. The idea is to generate a proper  `feature_levels.yaml` file reflecting the features and their relationships for each DB or tool that would generate specific GFF/GTF files. Then you always use the same template according to the origin of the file. A Research group generating a specific GFF/GTF file would share its AGAT `feature_levels.yaml` template along with the annotation in order to make the analyze reproducible.

> One such example is the [SGD reference GFF](http://sgd-archive.yeastgenome.org/sequence/S288C_reference/genome_releases/), which has a combination of RNA and non-RNA features, self-referencing feature types, feature types with multiple parent types, or leaf feature types pointing directly to root feature types ([here](https://drive.google.com/file/d/1oItDVksOw0rpq7PIWri_gzdSwveRlRpz/view?usp=sharing) is the corresponding graph).
> 
> Do you have a recommended way to automate the process of generating a good `feature_levels.yaml` file or is it something on the roadmap?

There is currently no automated way. I would run AGAT once to see what it says (it will list/inform what are the feature types contained within the file and tell you which one are missing in the `feature_levels.yaml`).  
I really like the idea to automatise this. I would be definitly feasible when ID/Parent relationship exists. But with GFF/GTF formated files using transcript_id/gene_id shared features and those using common attribute it would be more prone to error. The thing is it would definitly work if the file is correctly formated. But this is a nightmare, there are so many files with errors even in widely used DB...

**Reply by svigneau** [reply]
Thank you for your detailed answer!
